core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,269 @@
+use std::{any::Any, cell::RefCell, collections::HashMap, fmt};
+
+use futures::{future::BoxFuture, FutureExt};
+use tokio::{runtime::Runtime, sync::watch};
+use zksync_health_check::CheckHealth;
+
+pub use self::{context::NodeContext, stop_receiver::StopReceiver};
+use crate::{
+    resource::ResourceProvider,
+    task::{TaskInitError, ZkSyncTask},
+};
+
+mod context;
+mod stop_receiver;
+
+type TaskConstructor =
+    Box<dyn FnOnce(&NodeContext<'_>) -> Result<Box<dyn ZkSyncTask>, TaskInitError>>;
+type HealthCheckTaskConstructor = Box<
+    dyn FnOnce(
+        &NodeContext<'_>,
+        Vec<Box<dyn CheckHealth>>,
+    ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>,
+>;
+
+/// "Manager" class of the node. Collects all the resources and tasks,
+/// then runs tasks until completion.
+///
+/// Initialization flow:
+/// - Node instance is created with access to the resource provider.
+/// - Task constructors are added to the node. At this step, tasks are not created yet.
+/// - Optionally, a healthcheck task constructor is also added.
+/// - Once the `run` method is invoked, node
+///   - attempts to create every task. If there are no tasks, or at least task
+///     constructor fails, the node will return an error.
+///   - initializes the healthcheck task if it's provided.
+///   - waits for any of the tasks to finish.
```

Comment:
what about one-off tasks? We don't need support for those?

Reply:
IIRC whenever we use one-off tasks, we still want the node to exit upon completion, so it's kind of covered (e.g. task runs, returns `Ok(())`, and the node exits).
I don't think I'm aware of a use-case where we have a task that can complete but shouldn't cause the node to exit, do you have an example?

Reply:
we have this for some reason: https://github.com/matter-labs/zksync-era/blob/3a07890dacebf6179636c44d7cce1afd21ab49eb/core/bin/zksync_server/src/main.rs#L162

Reply:
Yep, but this was a kludge to implement one-off tasks.
_By default_ no actor is expected to exit, and if something exited, then depending on this boolean we will either exit with zero or non-zero code. I can't say I like this approach, so now tasks are allowed (but not required) to exit, and if task returns `Ok(())` this would be considered equivalent to `tasks_allowed_to_finish = true`.